### PR TITLE
Feature/charts

### PR DESCRIPTION
**Pull request by richavyas**
## Description

PR for Bar & Line Charts

### What's included?

- Single Line Chart
- Multi Line Chart
- Bar Chart
- Responsive Charts
- Supports three formats: `json`, `csv`, `tsv`
- Default `Color Palette` for Bar Chart
- Drop Shadow in Bar Chart
- `Config Parameters` : Axis Titles, Chart Title, dataSrc for data source, contentType, zDepth, lineTitles & lineColors for Multi Line Chart

#### Test Steps

- Navigate to http://localhost:4200/#/components/bar-chart & http://localhost:4200/#/components/line-chart
- Check charts

### What's coming next
- [x] creating custom color palette using two colors provided by user
- [x] changing nomenclature for all the charts and bringing them under one td-chart element
- [x] data limit
- [x] line/bars shown on the page limit
- [x] more configs

##### Screenshots or link to CodePen/Plunker/JSfiddle
covalent-charts

### CODE REVIEW

File: src/platform/charts/bar-chart/bar-chart.component.scss [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  }
+  .svg-content-responsive {
+      display: inline-block;
+      position: absolute;
+      top: 10px;
+      left: 0;
+  }
+  #z-depth5 {
+    height:130%;
+    feGaussianBlur {
```
Comment (kyleledbetter) [inline src/platform/charts/bar-chart/bar-chart.component.scss]: whoa you can put the svg blue into css like this? didn't know that!

File: src/platform/charts/bar-chart/bar-chart.component.scss [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    vertical-align: top;
+    overflow: hidden;
+  }
+  .svg-content-responsive {
+      display: inline-block;
+      position: absolute;
+      top: 10px;
+      left: 0;
+  }
+  #z-depth5 {
```
Comment (kyleledbetter) [inline src/platform/charts/bar-chart/bar-chart.component.scss]: do these need to be ideas or can you put a class on the chart (which can also have an ID if it still needs one)

**Comment by kyleledbetter** [comment]
a big thing i'm still waiting to see is splitting out the parent `<td-charts>` component w/ all its configs, and then a separate chart component for each series as a child

File: src/platform/charts/bar-chart/bar-chart.component.scss (hunk 2 above)
Comment (richavyas) [inline src/platform/charts/bar-chart/bar-chart.component.scss]: I added these to discuss but this was actually not adding any styles. another thing is that it's different elements where the styles need to be added so adding a class/id is probably not gonna work. That's why I added these as part of config eventually but left it hear for debate/thoughts.

**Comment by richavyas** [comment]
Yes, that's next, mentioned in `What's coming Next` section.

**Comment by richavyas** [comment]
- [X] td-chart => td-charts
- [X] td-bar-chart => td-chart-bar
- [X] td-line-chart => td-chart-line
- [X] custom color palette added; core color palettes added as palette options;
- [X] defualt palette for bar chart & error handling in case of unmatched color
- [X] grid/drop-shadow added to td-charts configuration and passed to individual charts

**Comment by kyleledbetter** [comment]
Very nice updates! Just glancing at the code a couple requests:

- [ ] can we move the duplicated core chart styles out of bar-chart.component.scss & line-chart.component.scss into a /charts/charts.component.scss (and delete all styles that aren't bar-specific and line-specific styles from those respective stylesheets)
- [ ] rename all the bar-chart directories and files to chart-bar to avoid confusion
- [ ] rename all the line-chart directories and files to chart-line

I can help w/ the scss stuff if you need

**Comment by richavyas** [comment]
Covered the above points in this commit

File: src/app/components/components/charts/charts.component.html [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+                     [zDepthConfig]="['125%', 2, 0, 2]"
+                     shadowColor="rgba(0, 0, 0, 0.54)"
+                     fillOpacity="1"
+                     chartTitle="Sales Bar Chart"
+                     [ticks]='ticksFlag'
+                     [grid]='gridFlag'
+                     [palette]='colorPalette'>
+      </td-chart-bar>
+      <md-divider></md-divider>
+      <md-card-title>Single/Multi Line Chart</md-card-title>
```
Comment (kyleledbetter) [inline src/app/components/components/charts/charts.component.html]: hmmm it feels weird to have non-chart html elements in a `<td-charts>` parent, I'd expect each series `<td-chart-bar>` and `<td-chart-line>` to be a bar chart and line chart on 1 td-chart, not 2 separate charts

**Comment by kyleledbetter** [comment]
so in moving things out to the `<td-chart>` parent i'd expect all the axis, grids, and all other general chart logic EXCEPT the series (like a line or bars) to be parameters and options of the parent/wrapper `<td-chart>`

something like:

```
<td-charts
            chartTitle="Sales Bar & Temperature Trends"
            leftAxisTitle="Sales"
            rightAxisTitle="Temperature"
            bottomAxisTitle="Day Offset"
            [zDepthConfig]="['125%', 2, 0, 2]"
            shadowColor="rgba(0, 0, 0, 0.54)"
            fillOpacity="1"
            [ticks]='ticksFlag'
            [grid]='gridFlag'>

      <td-chart-bar [palette]='colorPalette' barColumns="sales"  dataSrc="platform/charts/chart-bar/data.csv" contentType="csv"></td-chart-bar>

      <td-chart-line dataSrc="platform/charts/chart-line/data.tsv"
                   contentType="tsv"
                   bottomAxis="col1"
                   [lineColumns]="['col2', 'col3','col4']"
                   [lineTitles]="['New York','San Francisco', 'Austin']"
                   [lineColors]="['#3366cc', '#dc3912', '#ff9900']"></td-chart-line>

    </td-charts>
```

**Comment by kyleledbetter** [comment]
it would be nice if
`[palette]='colorPalette'`
and
`[lineColors]="['#3366cc', '#dc3912', '#ff9900']"`
had a similar name, like both could be `[colors]` that would accept a variable or an array like

`[colors]='colorPalette'`
and
`[colors]="['#3366cc', '#dc3912', '#ff9900']"`

**Comment by kyleledbetter** [comment]
i also thing the depth and color of the shadow should prob have similar names so you know they're related like:
instead of:
```
            [zDepthConfig]="['125%', 2, 0, 2]"
            shadowColor="rgba(0, 0, 0, 0.54)"
```
do
```
            [shadowDepth]="['125%', 2, 0, 2]"
            shadowColor="rgba(0, 0, 0, 0.54)"
```

**Comment by richavyas** [comment]
- [X] Migrated following chart params from individual chart selectors to `td-charts`:
- chartTitle
- leftAxisTitle
- rightAxisTitle
- bottomAxisTitle
- [zDepthConfig]
- shadowColor
- fillOpacity
- ticks
- grid

**Comment by richavyas** [comment]
This commit doesn't include combining two charts into one. The way charts are implemented is each selector like `td-chart-bar` & `td-chart-line` create separate svgs. So including a selector in html will create a new svg. Lets discuss this in detail.

**Comment by richavyas** [comment]
[X] sass fixed after rc.7 update
[X] Transitions and sorting added for bar chart
[X] architectural changes for creating svg, applying generic styles in `td-charts` component and actual chart creation in individual chart components like `td-chart-bar`/`td-chart-line`

bar-chart-transitions

**Comment by richavyas** [comment]
What's Next:
- [ ] Transitions for Line Chart
- [ ] Area Chart

**Comment by kyleledbetter** [comment]
Great updates! Here's my thoughts while testing!

Order of Grid VS Chart bars/lines/etc:

If you can append the grid & ticks *before* the chart bars/lines, it will render in the right order and you won't see the grid on top like it is currently:

image

so if you can render in this order it will fix it:

```
<g class="grid">...
<g class="ticks">...
<rect>...
```

also it would be nice if all the chart shapes (rect in this example) would be wrapped in a container like 

```
<g class="grid">...
<g class="ticks">...
<g class="chart chart-bars">
    <rect>...
```

**Comment by kyleledbetter** [comment]
## Two charts of the same type in one page

Currently it seems like if I try to have two separate line charts in 1 page, they both get appended to the same element instead of rendering two charts:

my test code:
```
    <md-card-title>Line Chart One</md-card-title>
    <td-charts chartTitle="Temperature Trend 2"
                bottomAxisTitle="Day Offset"
                leftAxisTitle="Temperature"
                [grid]="true"
                [ticks]="true"
                [shadowDepth]="['125%', 2, 0, 2]"
                shadowColor="rgba(0, 0, 0, 0.54)"
                fillOpacity="1">
      <td-chart-line 
                  dataSrc="platform/charts/chart-area/data.json"
                   contentType="json"
                   bottomAxis="col1"
                   [lineColumns]="['col2']"
                   [lineTitles]="['Cali']"
                   colors="colorPalette">
      </td-chart-line>
    </td-charts>
    <md-divider></md-divider>
    <md-card-title>Line Chart Two</md-card-title>
    <td-charts chartTitle="Temperature Trend"
                bottomAxisTitle="Day Offset"
                leftAxisTitle="Temperature"
                [grid]="true"
                [ticks]="true"
                [shadowDepth]="['125%', 2, 0, 2]"
                shadowColor="rgba(0, 0, 0, 0.54)"
                fillOpacity="1">
      <td-chart-line dataSrc="platform/charts/chart-line/data.tsv"
                   contentType="tsv"
                   bottomAxis="col1"
                   [lineColumns]="['col2', 'col3','col4']"
                   [lineTitles]="['New York','San Francisco', 'Austin']"
                   [colors]="['#3366cc', '#dc3912', '#ff9900']">
      </td-chart-line>
    </td-charts>
```

image

**Comment by kyleledbetter** [comment]
## Content Type attribute

Do we need to call this out specifically? `contentType="tsv"` or can we just get the filename extension and use that `dataSrc="platform/charts/chart-line/data.tsv"` it would be nice to not have to call it out

**Comment by kyleledbetter** [comment]
## Grid Options

instead of just having 1 boolean `[grid]="true"` it would be nice to have options for either grid, and also possibly the color:
```
[gridX]="true"
gridXColor="rgba(0, 0, 0, 0.54)"
[gridY]="false"
gridYColor="rgba(0, 0, 0, 0.54)"
```

**Comment by kyleledbetter** [comment]
## Transitions

Obviously I ❤️ the transitions. It would great to have options. I think the options would go on each child chart component like:

at the very least we need a boolean to turn them on or off
```
<td-chart-line [transition]="true">
```
but it would be really nice to have options like:
```
<td-chart-line 
                   [transition]="true"
                   transitionDuration="3000"
                   transitionType="easeLinear"
                   transitionDelay="1000">
```

**Comment by kyleledbetter** [comment]
## Different attribute names

So from chart to chart, some similar attributes have different names like `barColumns` vs `lineColumns` vs `areaColumns` and `lineTitles`, etc

Could those not just be `columns` and `titles` or something more consistent from chart to chart?

**Comment by richavyas** [comment]
- [X] support for multiple similar charts
- [X] changes for DOM e.g. actual chart elements inside a 'g' 
   ```<g class="chart-bars">
    <rect>...```
- [X] more reusability
- [X] parameters name changes for all the charts e.g. `lineColumns` => `columns`

**Comment by richavyas** [comment]
What's Next:
- [ ] More Error Handling
- [ ] Transition enhancements

**Comment by kyleledbetter** [comment]
tested latest and they look great! I pushed a small change to get the area chart on top of the grid also

next up:

- [ ] use material color palettes (like `[colors]="['cyan', 'indigo']"`) on areas and lines
- [ ] get shadow options working on area chart
- [ ] tick labels are now under the chart (since we changed order) so either split them out or render them outside the axis instead of inside, here's the issue:
image

and to keep track of recents:

- [ ] get content type from filename instead of requiring contentType="tsv"

**Comment by richavyas** [comment]
Added:
- [X] Line Chart Transitions
- [X] Transition options like 
`<td-chart-line 
                   [transition]="true"
                   transitionDuration="3000"
                   transitionDelay="1000">`
- [X] Error handling for wrong template colors

line-chart-transitions

screen shot 2016-10-04 at 12 04 27 am

**Comment by richavyas** [comment]
What's next:
- [ ] Handle time series based data

**Comment by kyleledbetter** [comment]
Great work on the latest! Tested and came across some requests:

- [ ] ability to hide the axis (something like showLeftAxis="false" and showBottomAxis="false" to hide them)
- [ ] chartHeight parameter for td-chart, so you can set the height but the width remains fluid

and just a reminder for previous requests:
- [ ] use material color palettes (like `[colors]="['cyan', 'indigo']"`) on areas and lines
- [ ] get shadow options working on area chart
- [ ] tick labels are now under the chart (since we changed order) so either split them out or render them outside the axis instead of inside
- [ ] get content type from filename instead of requiring contentType="tsv"

**Comment by kyleledbetter** [comment]
also it seems like fillOpacity has some issues. When i edit the fillOpacity on the area chart, it affect the other two charts and not itself :)

image

**Comment by richavyas** [comment]
Includes:
- Time Series for Line Chart
- [X] ability to hide the axis (something like showLeftAxis="false" and showBottomAxis="false" to hide them)
- [X] get shadow options working on area chart
- [X] tick labels are now under the chart (since we changed order) so either split them out or render them outside the axis instead of inside
 - [X] get content type from filename instead of requiring contentType="tsv"

**Comment by richavyas** [comment]
Things to be added:
- [ ] minor changes in all the charts for Time series support
- [ ] opacity fix
- [ ] chart height
- [ ] palette for line and chart (requires discussion)

**Comment by kyleledbetter** [comment]
All updates look great!

Latest requests:

- [ ] bottomAxisTitle doesn't appear to be working
- [ ] basic tooltips would 🎸 
- [ ] ability to hide each of the grid lines individually
- [ ] when showBottomAxis & showLeftAxis the left and bottom line should also be hidden; example:
image

**Comment by emoralesb05** [comment]
-  Removed `d3` from `scripts:[]` in `angular-cli.json` and favored `require()` since that way avoid configuration on consumption.
- Moved charts menu-item and icon closer to the `viz` components in menus.

**Comment by kyleledbetter** [comment]
I pushed some updates for theme & scss and noticed:

- [ ] opacity still affects all charts on the page
- [ ] hiding individual grids setting still needed

**Comment by emoralesb05** [comment]
So i noticed a couple of things:

## Rendering

- Charts depend on a file at the moment which is nice to have, but i was expecting a `data` input for an object since that is what we normally get in `listener` or from API's.
- Really need to think over how we are gonna merge 2 child charts into the same container since listener has that use case.

## Refresh

- We need a way to refresh and update the chart when `data` (when added) or `dataSrc` changes + a way to programatically add a point or w/e referencing the component.
e.g. 
```
@ViewChild(TdChartsComponent) chart: TdChartsComponent;
 
  ngAfterViewInit() {
    this.chart.add(/*point*/);
    this.chart.update();
  } 
```

**Comment by emoralesb05** [comment]
I found a bug in `area-chart` where the chart content has an offset and its a little bit to the left.

image

Its supposed to start at 0 and end at 500 in this example.

**Comment by emoralesb05** [comment]
Because of the `transitionAndSorting` method in bar chart, x labels and bars dont match.

image

**Comment by emoralesb05** [comment]
- [ ] Refactored code into an abstract `ChartComponent` class so its more maintainable and easier to add additional charts so we only need to focus on the chart concrete implementation.
- [ ] Added `[data]` input for object rendering.
- [ ] Added validation if no `data` or `dataSrc` are added.
- [ ] Bug fix for area chart where there was an offset.
- [ ] Set default opacity to 1 and default height to 450.

**Comment by emoralesb05** [comment]
- [ ] Changed code so charts are rendered into the same container (parent) instead of different ones so we can later implement multiple charts.

**Comment by emoralesb05** [comment]
- [ ] Removed sorting from `bars` for now.
- [ ] Added axis components for x, y-left and y-right that get linked with charts to be able to render 2 charts at the same time correctly.
e.g.
```
<td-charts title="Sales Bar/Line Chart"
              [grid]="true"
              [shadowDepth]="['125%', 2, 0, 2]"
              shadowColor="rgba(0, 0, 0, 0.54)"
              fillOpacity="0.95">
      <td-axis-x [link]="chartLine4" [ticks]="true" [show]="true" legend="Day Offset"></td-axis-x>
      <td-axis-y-left [link]="chartLine4" [ticks]="true" [show]="true" legend="Sales"></td-axis-y-left>
      <td-axis-y-right [link]="chartBar2" [ticks]="true" [show]="true" legend="Sales 2"></td-axis-y-right>
      <td-chart-bar #chartBar2 [colors]="['orange', 'deep-orange']"
                    dataSrc="platform/charts/chart-line/datatime.csv"
                    bottomAxis="date"
                    columns="close"
                    titles="close"
                    [transition]="true"
                    transitionDuration="750"
                    transitionDelay="50">
      </td-chart-bar>
      <td-chart-line #chartLine4 dataSrc="platform/charts/chart-line/datatime.csv"
                      timeSeries="true"
                      bottomAxis="date"
                      [columns]="['close']"
                      [titles]="['close']"
                      [colors]="['#03a9f4']"
                      [transition]="true"
                      transitionDuration="500"
                      transitionDelay="1000">
      </td-chart-line>
    </td-charts>
```

output:
image

**Comment by kyleledbetter** [comment]
Ok last few updates before merging!

- [ ] currently the grid lines & ticks are both wrapped with `class="grid"` and each child is `class="tick"`, the ticks are the tiny lines by the numbers outside the chart, the grid lines are the horizontal or vertical lines in the grid itself, need to separate these and have attr to hide any/all of them individually
- [ ] params on `td-charts` to set the margin whether in an array `<td-charts [margin]="[10,20,50,20]">` (that's top,right,left,bottom) or individually `<td-charts marginTop="100"`
- [ ] ability to set `fillOpacity` on each individual chart
- [ ]  small thing but the td-chart-line always has this little starting line before it gets to the values:
image
- [ ] multi-lines have this gap on the left and right of the lines
image



We're close!

**Comment by richavyas** [comment]
- [X] Default padding and padding as parameter added in `<td-chart-bar>`

**Comment by richavyas** [comment]
- [X] Modified Bar chart transitions to go from bottom to top

bar-chart-transitions-mod

combo-chart-transitions-mod

**Pull request merged**